docs/error_handling.md

```diff
@@ -0,0 +1,200 @@
+
+The returned distinct type will be useless at the call-site unless
+it is stripped-away through `raising`, `either` or `check` which are
+the error-handling mechanisms provided by this library and discussed
+further in this document.
+
+If you accidentally forget to use one of the error-handling mechanisms,
+you'll get a compilation error along these lines:
+
+```
+required type for x: float
+  but expression 'Raising[(ValueError, KeyError), float](foo_original(x, y))' is of type: Raising[tuple of (ValueError, KeyError), system.float]
+```
+
+Please note that if you have assigned the `Raising` result to a
+variable, the compilation error might happen on a line where you
+attempt to use that variable. To fix the error, please introduce
+error handling as early as possible at the right call-site such
+that no `Raising` variable is created at all.
+
+`noerrors` is another pragma provided for convenience which is
+equivalent to an empty `errors` pragma. The forced error handling
+through the `Raising` type won't be applied.
+
+Both pragmas can be combined with the `nodefects` pragma that
+indicates that the specific proc should be proven to be Defect-free.
+
+The transformation uses a template by default to promote efficiency,
+but if you need to take the address the Raising proc, please add the
+`addressable` pragma that will force the wrapper to be a regular proc.
+
+Finally, `failing` is another pragma provided for convenience which
+is equivalent to `{.errors: (CatchableError).}`.
+
+## The `raising` annotation
+
+The `raising` annotation is the simplest form of error tracking
+similar to the `try` annotation made popular by the [Midori error model](http://joeduffyblog.com/2016/02/07/the-error-model/),
+which is also [proposed for addition in C++](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r0.pdf) and [already available in Zig](https://ziglang.org/documentation/master/#try).
+
+It merely marks the locations in the code where errors might be raised
+and strips away the `Raising` type to disarm the compiler checks:
+
+```nim
+proc attachValidator(node: BeaconNode, keyFile: string) {.failing.} =
+  node.addValidator(raising ValidatorPrivKey.init(raising readFile(keyFile))
+```
+
+When applied to a `Result` or an `Option`, `raising` will use the
+`tryGet` API to attempt to obtain the computed value.
+
+## The capital `Try` API
+
+The capital `Try` API is similar to a regular `try` expression
+or a `try` statement. The only difference is that you must provide
+exception handlers for all possible recoverable errors. If you fail
+to do so, the compiler will point out the line in the `try` block
+where an unhandled exception might be raised:
+
+```nim
+proc replaceValues(values: var openarray[string],
+                   replacements: Table[MyEnum, string]) =
+  Try:
+    for v in mitems(values):
+      let
+        enumValue = parseEnum v
+        replacement = replacements[enumValue] # Error here
```

Comment:
this is the hallmark of exception-unsafe code, and a school-book example of a design issue in `Try`: mid-way through the iteration on values, an exception is raised and you're left with a partially mutated `values` input which the calling code _cannot_ reason about and that is invisible when reading the code, unless you carefully analyze the control flow - there's no visual indication for the reader that this_ in particular is where there `KeyError` happens.

The most "natural" thing to do if the compiler is pestering me about `KeyError` here is to add a `except KeyError` below `except ValueError` which hides the actual exception safety issue with this code.

Reply:
notably, the `raising` annotation does not suffer from this issue, as I understand it

Reply:
Well, this is just an illustrative example. After all, the code doesn't compile, so it being a "school-book example of a design issue" is reasonable. If you can offer a better example with exceptions that every Nim user will be familiar with, feel free to suggest it.

Reply:
do you need to put `raising` inside `Try` or not?

Reply:
You have to put it in APIs that use the `errors` pragma.